csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -8,46 +8,314 @@
 #include <debug.h>
 #include <ir/utils.h>
 #include <logical_domain_map.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
 #include <options.h>
 #include <preseg_passes/remove_bcast_squeeze.h>
 
 namespace nvfuser::preseg_passes {
 
 namespace {
-// Remove broadcast-squeeze and squeeze-broadcast patterns
-// TODO: still remove when have intermediate ops between broadcast and squeeze
-void removeBcastSqueeze(Fusion* fusion) {
-  // Iterate backwards over fusion expressions.
-  // This will ensure we don't process expressions that are no longer valid
-  // after replacement.
-  auto exprs = fusion->exprs();
-  for (auto it = std::rbegin(exprs); it != std::rend(exprs); it++) {
-    Expr* expr = *it;
-    // step-1: find and remove broadcast + squeeze pattern
-    // before: Y = broadcast(X); Z = squeeze(Y);  M = someOp(Z)
-    // after : M = someOp(X)
-    // conditions: (1) broadcast and squeeze have the same dim flags
-    if (auto squeeze = dynamic_cast<SqueezeOp*>(expr)) {
-      if (auto bcast =
-              dynamic_cast<BroadcastOp*>(squeeze->in()->definition())) {
-        if (bcast->getBroadcastDimFlags() == squeeze->getSqueezeDimFlags()) {
-          ir_utils::replaceValInAllExprInputsAndFusionOutputs(
-              squeeze->out(), bcast->in());
-        }
+
+enum class AxisOp { PRESERVE, SQUEEZE, BROADCAST };
+
+std::ostream& operator<<(std::ostream& os, AxisOp op) {
+  switch (op) {
+    case AxisOp::PRESERVE:
+      os << "PRESERVE";
+      break;
+    case AxisOp::SQUEEZE:
+      os << "SQUEEZE";
+      break;
+    case AxisOp::BROADCAST:
+      os << "BROADCAST";
+      break;
+  }
+  return os;
+}
+
+//! This represents the combined state of a collection of broadcast and squeeze
+//! operations
+using AxisOps = std::vector<AxisOp>;
+
+//! Convert a broadcast Expr to an AxisOps descriptor
+AxisOps broadcastToAxisOps(BroadcastOp* bcast) {
+  AxisOps ops;
+  const std::vector<bool>& flags = bcast->getBroadcastDimFlags();
+  ops.reserve(flags.size());
+  for (bool flag : flags) {
+    ops.push_back(flag ? AxisOp::BROADCAST : AxisOp::PRESERVE);
+  }
+  return ops;
+}
+
+//! Convert a squeeze Expr to an AxisOps descriptor
+AxisOps squeezeToAxisOps(SqueezeOp* squeeze) {
+  AxisOps ops;
+  const std::vector<bool>& flags = squeeze->getSqueezeDimFlags();
+  ops.reserve(flags.size());
+  for (bool flag : flags) {
+    ops.push_back(flag ? AxisOp::SQUEEZE : AxisOp::PRESERVE);
+  }
+  return ops;
+}
+
+//! Check whether the loop domains are the same size and have the same
+//! parallelization, ignoring reduction dimensions.
+bool hasCompatibleParallelization(TensorView* orig_tv, TensorView* new_tv) {
```

Comment:
There's a utility called `isResharding` that you can leverage.

Reply:
Awesome. Thanks for pointing me to this.